### PR TITLE
Only allow positive values in viewportWidth/Height, use documentElement

### PR DESCRIPTION
**Pull request by jonsnyder**
## Description

Use window.document.documentElement.clientWidth/Height instead of window.innerWidth/Height. Add a check for negative values. If the values are negative or not present, don't include the data in the final xdm.

## Related Issue

https://jira.corp.adobe.com/browse/PDCL-4786

## Motivation and Context

Some customers are getting negative width or height values causing XDM validation errors. https://stackoverflow.com/questions/4629969/ios-return-bad-value-for-window-innerheight-width/39345914

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [x] I have made any necessary test changes and all tests pass.
- [ ] I have run the Sandbox successfully.

### CODE REVIEW

File: src/components/Context/injectEnvironment.js [hunk 1]
```diff
@@ -14,14 +14,24 @@ import { deepAssign } from "../../utils";
...
+
     const environment = {
-      type: "browser",
-      browserDetails: {
-        viewportWidth: innerWidth,
-        viewportHeight: innerHeight
-      }
+      type: "browser"
     };
+    const hasWidth = clientWidth && clientWidth >= 0;
```
Comment (Aaronius) [inline src/components/Context/injectEnvironment.js:24]: This probably isn't what you're intending. If `clientWidth` is `0`, then `hasWidth` will end up being `0`, which means `environment.browserDetails.viewportWidth` won't be set. It's probably worth adding a unit test for this case.

**Review by Aaronius — CHANGES_REQUESTED** [review]
Overall it's good. Just need to fix the `0` value issue.

File: src/components/Context/injectEnvironment.js (hunk 1 above)
Comment (jonsnyder) [reply src/components/Context/injectEnvironment.js:24]: Thanks for finding this!

File: src/components/Context/injectEnvironment.js [hunk 2]
```diff
@@ -10,18 +10,28 @@ OF ANY KIND, either express or implied. See the License for the specific languag
...
 */
 
-import { deepAssign } from "../../utils";
+import { deepAssign, isNumber } from "../../utils";
 
 export default window => {
   return xdm => {
-    const { innerWidth, innerHeight } = window;
+    const {
+      document: { documentElement: { clientWidth, clientHeight } = {} }
```
Comment (liljenback) [inline src/components/Context/injectEnvironment.js:18]: So AppMeasurement is using offsetWidth here. What made you pick clientWidth? There seems to be a little bit of a difference in that offsetWidth includes some additional browser elements etc.

File: src/components/Context/injectEnvironment.js [hunk 3]
```diff
@@ -10,18 +10,28 @@ OF ANY KIND, either express or implied. See the License for the specific languag
...
-      type: "browser",
-      browserDetails: {
-        viewportWidth: innerWidth,
-        viewportHeight: innerHeight
-      }
+      type: "browser"
     };
+    const hasWidth = isNumber(clientWidth) && clientWidth >= 0;
+    const hasHeight = isNumber(clientHeight) && clientHeight >= 0;
+    if (hasWidth || hasHeight) {
```
Comment (liljenback) [inline src/components/Context/injectEnvironment.js:26]: Is there a valid use-case for not doing && here?

File: src/components/Context/injectEnvironment.js (hunk 2 above)
Comment (jonsnyder) [reply src/components/Context/injectEnvironment.js:18]: I was just following the stack overflow answer linked above. I didn't realize appMeasurement was using documentElement as a fallback. offsetWidth and offsetHeight are probably better.

File: src/components/Context/injectEnvironment.js (hunk 3 above)
Comment (jonsnyder) [reply src/components/Context/injectEnvironment.js:26]: I can't think of one. I like it.

File: src/components/Context/injectEnvironment.js (hunk 2 above)
Comment (liljenback) [reply src/components/Context/injectEnvironment.js:18]: Using the property that will fluctuate the least (based on margin settings etc.) to enable for common resolutions to group well would probably be preferred.

**Review by jfkhoury — APPROVED** [review]

**Review by liljenback — APPROVED** [review]

**Pull request merged**